c_src/erl_optics.c

```diff
@@ -46,14 +48,49 @@ static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
     atom_dist = enif_make_atom(env, "dist");
     atom_histo = enif_make_atom(env, "histo");
     atom_quantile = enif_make_atom(env, "quantile");
-    poller = optics_poller_alloc();
-    ctx = enif_alloc(sizeof(struct map_ctx));
-    if (poller) {
-      optics_poller_backend(poller, (void *) ctx, backend_eo, backend_free_eo);
-    }
+    carbon_poller = 0;
+    erlang_poller = 0;
     return 0;
 }
 
+
+static ERL_NIF_TERM eo_allocate_carbon_poller(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+    if (!poller){
+        poller = optics_poller_alloc();
+    }
+    if (!poller) return ERROR("poller_allocation_error");
+    if (carbon_poller == 0) {
+        size_t host_size = 255;
+        size_t port_size = 255;
+        char host[host_size];
+        char port[port_size];
```

Comment:
Currently you're using a dynamic size of the array when you could easily use a static size here. The difference is that one requires a manual increment of the stack during the execution of the code while the other will just happen automatically in the preamble of the function. It's also a good idea to keep constant values constant.

There's a couple of ways to fix this, but the sure-est fire way to have a compile-time value is to do it through enums:
```
enum {host_size = 255, port_size = 255 };
char host[host_size] = {0};
char port[port_size] = {0};
```
It looks a bit weird but enums are basically guaranteed to be compile time value which is basically the only option in C without resorting to macros. Note that I think the more popular option out there is macros but I've made it my life's purpose to avoid macros wherever possible :)

Reply:
your port length is very generous :)